w07-ruby-on-rails-cont/d2-drills/README.md

```diff
@@ -0,0 +1,91 @@
+
+[![Divide and conquer (complex)](https://s3.amazonaws.com/ka-cs-algorithms/divide_conquer_3_steps.png)](https://www.khanacademy.org/computing/computer-science/algorithms/merge-sort/a/divide-and-conquer-algorithms)
+
+## Sorting Things
+
+If you were given an array of 10 random numbers, how would you sort them in ascending order?
+
+Arrays in Ruby have a [`#sort`](http://ruby-doc.org/core-2.2.0/Array.html#method-i-sort) method. If we have access to a Ruby runtime we can treat this as a small problem and simply call the built-in `Array#sort` method.
+
+Without a Ruby runtime, we'd be on our own. We'd have to think of sorting arrays as a big problem and come up with an implementation ourselves.
+
+## Algorithm Design
+
+Sorting algorithms might seem like a contrived example because most languages have built-in functionality for sorting. When building web apps, we generally avoid re-inventing the wheel by reusing functionality that the language or framework provides.
+
+Sorting algorithms are a popular interview topic because they force candidates to take a very disciplined approach. The seemingly simple problem of sorting an array of integers quickly becomes complex when you don't have access to `Array#sort`. Just explaining the steps involved can give the interviewer a lot of insight into your abilities as a problem solver.
+
+## Sorting Recursively
+
+There are many ways to sort an array of numbers. Each have their pros and cons in terms of time complexity, space complexity, readability, etc.
+
+Today, we'll focus on a recursive solution called merge sort.
+
+### Why Merge Sort?
+Merge sort is the first powerful sorting algorithm that you will encounter in the wilds of the real world (baked into Safari and Firefox.)  It uses an extremely efficient application of the 'Divide and Conquer' concept to lists of elements.  We worked on Bubble Sort previously, now let's up our game and work on Merge Sort!
+
+Merge Sort works on the basic principal of dividing your list into sub-lists (recursively) until your sub-lists are of length one or zero.  Once your sub-lists are at that size, you merge with a neighboring sub-list.  When you merge them, you merge them in ascending or descending order, depending on your implementation.  
+
+![Merge Sort visualization](https://webdocs.cs.ualberta.ca/~holte/T26/Lecture6Fig6.gif)
+
+
+## Challenge: Write your own Merge Sort
+
+There are TWO functions that work together to accomplish a Merge Sort:
+
+-  A `merge_sort` function that:
+  - takes an array as a parameter
+  - splits the input array into two subarrays
+  - calls a `merge` function  
+
+*Note:* The `merge_sort` function **is recursive**. Don't forget your base case!
+
+-  A `merge` function that:
```

Comment:
You're right about this. Should definitely add some code snippets to get them started.
We need to specify that the `merge` function takes two arrays (`left` and `right`) as arguments, as well as more explicitly explaining the "split input array into subarrays" part. Something like "create two subarrays from the input array using the `Array#[] method` combined with the `..` and `...` range operators."

`def merge_sort(array)`

`left = [something...something else]` 
`right = [something else .. end]`

`merge(left, right)`
`results = []`
`if left.first <= right.first`
`results << left.first`
`else`
`results << right.first`

`results + left + right`
^This might need to be explained to them with some sample code in IRB (concatenating arrays).